src/workerd/api/tracing-module.c++

```diff
@@ -0,0 +1,225 @@
+      }
+      KJ_CASE_ONEOF(str, kj::String) {
+        s.setTag(kj::mv(key), kj::mv(str));
+      }
+    }
+  }
+}
+
+bool InternalSpanImpl::getIsRecording() {
+  return span != kj::none;
+}
+
+SpanParent InternalSpanImpl::makeSpanParent() {
+  KJ_IF_SOME(s, span) {
+    return SpanParent(s);
+  } else {
+    return SpanParent(nullptr);
+  }
+}
+
+SpanParent InternalSpan::makeSpanParent() {
+  KJ_SWITCH_ONEOF(impl) {
+    KJ_CASE_ONEOF(s, kj::Own<InternalSpanImpl>) {
+      return s->makeSpanParent();
+    }
+    KJ_CASE_ONEOF(s, IoOwn<InternalSpanImpl>) {
+      return s->makeSpanParent();
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+jsg::Ref<InternalSpan> TracingModule::startSpan(jsg::Lock& js, const kj::String name) {
+  auto spanName = kj::ConstString(kj::str(name));
+
+  if (IoContext::hasCurrent()) {
+    auto& ioContext = IoContext::current();
+    auto span = ioContext.makeUserTraceSpan(kj::mv(spanName));
+    auto impl = kj::heap<InternalSpanImpl>(kj::mv(span));
+    return js.alloc<InternalSpan>(ioContext.addObject(kj::mv(impl)));
+  } else {
+    // When no IoContext is available, create a null span
+    auto impl = kj::heap<InternalSpanImpl>(nullptr);
+    return js.alloc<InternalSpan>(kj::mv(impl));
+  }
+}
+
+v8::Local<v8::Value> TracingModule::startSpanWithCallback(jsg::Lock& js,
+    kj::String operationName,
+    v8::Local<v8::Function> callback,
+    jsg::Arguments<jsg::Value> args,
+    const jsg::TypeHandler<jsg::Ref<InternalSpan>>& jsSpanHandler,
+    const jsg::TypeHandler<jsg::Promise<jsg::Value>>& valuePromiseHandler) {
+
+  // Get the current span parent
+  SpanParent parent = SpanParent(nullptr);
+  if (IoContext::hasCurrent()) {
+    parent = IoContext::current().getCurrentTraceSpan();
+  }
+
+  // Create the span implementation
+  auto spanName = kj::ConstString(kj::str(operationName));
+
+  // Create the JavaScript span object
+  jsg::Ref<InternalSpan> jsSpan = nullptr;
+  if (IoContext::hasCurrent()) {
+    auto& ioContext = IoContext::current();
+    auto span = ioContext.makeUserTraceSpan(kj::mv(spanName));
+    auto impl = kj::heap<InternalSpanImpl>(kj::mv(span));
+    jsSpan = js.alloc<InternalSpan>(ioContext.addObject(kj::mv(impl)));
+  } else {
+    // When no IoContext is available, create a null span
+    auto impl = kj::heap<InternalSpanImpl>(nullptr);
+    jsSpan = js.alloc<InternalSpan>(kj::mv(impl));
+  }
+
+  // Create new span parent for the callback execution context
+  SpanParent newSpanParent = jsSpan->makeSpanParent();
+
+  // Prepare callback arguments
+  v8::LocalVector<v8::Value> argv(js.v8Isolate);
+  argv.push_back(jsSpanHandler.wrap(js, jsSpan.addRef()));
+  for (auto& arg: args) {
+    argv.push_back(arg.getHandle(js));
+  }
+
+  // Define callback execution logic
+  auto executeCallback = [&jsSpan, &js, &callback, &argv, &valuePromiseHandler]() -> v8::Local<v8::Value> {
+    auto v8Context = js.v8Context();
+    return js.tryCatch([&]() -> v8::Local<v8::Value> {
+      auto result =
```

Comment:
The `callback` really ought to be either a `jsg::Function<...>` or a `jsg::JsFunction`, either of which should make this logic a bit easier to implement